### ISSUE TITLE
Manage DF data in the cloud

### ISSUE THREAD

**Issue by myk002** [DFHack/dfhack#3969]
As a prerequisite for Cloud Gaming (https://github.com/DFHack/dfhack/issues/3574), DFHack would need to provide a mechanism for players to copy their DF savegame and config data to an alternate directory that is managed by DFHack's Steam Cloud. This is necessary because DF doesn't offer Cloud saving by itself.

We could either add a separate top-level directory (e.g. `dfhack-cloud/`) for cloud backup or use a subdirectory under `dfhack-config/`, which is already enabled for Steam Cloud.

To save/restore DF game settings, we would need to act before DF is started, which means implementing at least some functionality in dflaunch (or the `dfhack` script? what platform does cloud gaming run as?).

There could also be a plugin, which would provide an in-game enablement and management interface.

when the functionality is enabled, either the plugin or the launcher would rsync data from `save/` to the cloud save directory whenever DF saves the game. before game load, the launcher (must be the launcher since a plugin wouldn't be able to do it early enough) would rsync data from the cloud save directory to the DF directories.

**Comment by dcraig327** [comment]
Cloud gaming on GeForce NOW runs as Windows 11.

Envision a server with only one game installed, the game you choose to play, in a Windows 11 VM and the first thing you see is the familiar Steam App showing the Library with only that game purchased and loading in the background.

**Comment by dcraig327** [comment]
Personally I like `dfhack-cloud/`. So much so to rename `dfhack-config/` to `dfhack-cloud/` and adding the saves to `dfhack-cloud/save/`. This still allows the base game to enable Steam Cloud functionality in the future, also it's easy to understand for users and adds minimal clutter to the root directory. Yet, because I like it doesn't mean it's necessary or even worthwhile. In fact whatever location you choose to store the save games & config data will discovered and documented for all users to see on the PC Gaming Wiki page: https://www.pcgamingwiki.com/wiki/Dwarf_Fortress 

Steam does the Cloud Sync right before the game is launched and also immediately after the game exits. This `dfhack-cloud/save/` directory would be sync'd upon entering `launchdf.exe` and however you've hooked dfhack to be be notified right before the DF process terminates, as long as the saves are in this `dfhack-cloud/save/` before that happens, then it'll be uploaded to the cloud.

Yes, only sync to/from the cloud save directory when the user has, you may have worded this differently, enabled cloud functionality. Within Steam as you can disable it two ways, under the games properties and in Steam Settings > Cloud.

**Comment by lethosor** [comment]
I want to avoid renaming `dfhack-config` to `dfhack-cloud`. It's not appropriate for non-cloud players, who would likely be the majority for a while.

More generally, I think a plugin would be ok. I think it would need to be optional in order to (a) avoid unwanted bandwidth consumption and (b) avoid consuming Steam cloud storage (is there a limit? I thought that was why DF opted not to enable Steam Cloud for everyone...)

**Comment by myk002** [comment]
There are technical size limits, but they are large. My understanding about why DF does not sync savegames to the cloud is that Toady didn't want the players to experience the delay.

![image](https://github.com/DFHack/dfhack/assets/977482/2fda6f70-9d27-4843-9536-26e0886c5532)

I opened a support request to Steam asking for more information about what would be possible:
```
I have a question about how my app might function with Geforce NOW. DFHack (app id: 2346660) runs as an in-process addon to Dwarf Fortress (app id: 975370). DFHack and Dwarf Fortress install to the same directory, and DFHack is dynamically loaded by Dwarf Fortress if the files are present. When DFHack is run from the Steam client, our launcher process runs the Dwarf Fortress app.

If we make DFHack available to Geforce NOW players, can our launcher dynamically install Dwarf Fortress on the Geforce NOW VM if the player has purchased Dwarf Fortress?
```

They said:
```
Sorry, but you'll need to reach out to nVidia about this. Our support for GeForce NOW just resolves around us allowing apps to participate, but we're not the experts on how things really work on that platform.
```

I then wrote an email to the Geforce NOW team:
```
I have an app published on Steam: DFHack (app id: 2346660)
https://steamdb.info/app/2346660/

It runs as an addon to Dwarf Fortress (app id: 975370)
https://steamdb.info/app/975370/

DFHack and Dwarf Fortress install to the same directory, and DFHack is dynamically loaded by Dwarf Fortress if the binary files are present. When DFHack is run from the Steam client, the launcher process runs the Dwarf Fortress binary.

If we make DFHack available to Geforce NOW players, can our launcher dynamically install Dwarf Fortress on the Geforce NOW VM via the Steam API if the player has purchased Dwarf Fortress?

Also, does Dwarf Fortress have to explicitly support GeForce NOW itself, or can my app proxy that support, e.g. my managing cloud saves on behalf of the Dwarf Fortress app?
```

but I haven't seen a response yet.

**Comment by dcraig327** [comment]
I imagine that nVidia GFN doesn't allow Installing 2 games with different appid's (DF & dfhack). Technically they are the same game as they both install to the same directory, but within the Steam client it requires installing two different games. I do feel they should add this feature and your wording to nVidia is great.

Just tested SKSE (Skyrim Script Extender) as I own that and the original Skyrim on Steam. And after installing only SKSE and after clicking Play, I'm prompted to install Skyrim as if I had clicked the Install button on Skyrim. This could be mentioned to nVidia if they're stumped. It also could be added to dfhack. I did check SKSE and there's no mention of the Skyrim depot at all: https://steamdb.info/app/365720/depots/ 

I mentioned X3: Albion Prelude as an example of a "DLC" that installs the base game, but this DLC isn't mentioned in the base game's store page. I just checked that games depot and it has the entire listing for the base game X3: Terran Conflict there and the size of that depot matches the size of the base game's depot:
https://steamdb.info/app/201310/depots/
https://steamdb.info/app/2820/depots/

Regarding having to zip up the save folder to speed up cloud syncing. I imagine a low priority process could be created in the background handling that while the game is running. Low priority. Also, possibly compressing the majority of files that are least likely to change and flagging that archive dirty if one of those files do change. To me that sounds like a lot of work, setting up a database for each file almost. Yet I imagine if anyone really was worried about minimizing the overhead time of syncing saves to and from the cloud, better ideas would come. Logging a count of how many times a file was written to during a playthrough, just a file name and a number. That'd be an efficient way of creating a starting point of how the files would be organized in these archives. Having multiple archives would be preferred.

Compression would happen in the background on one thread assigned a lower priority while the game is running. Yet when the game isn't running. You can create multiple threads to use almost all of the CPU's capabilities and speed up the decompression and compression.

zip likely isn't the most efficient. 7z I imagine is the fastest. Found an open source GPL3 project that indicates it's based off of 7z and much faster: https://peazip.github.io/fastest-compression-zip-deflate-benchmark.html

**Comment by dcraig327** [comment]
Source to that old version of SKSE listed in Steam: https://github.com/NightQuest/SKSE/

**Comment by dcraig327** [comment]
> but I haven't seen a response yet.
I have a feeling it may remain this way. To me a a lack of a response past 24 hours means I get to get louder until there is a response. It could be some corporate excuse mentioning security, but a lack of a response is unprofessional.

> Control what you can control.
I say ignore Cloud Gaming at the moment as this seems to be held up by means outside of your control.

I say add DF saves to DFHack's Steam Cloud to enable proper Steam Cloud support for DF outside of Cloud Gaming. People would love it.

You did what you can. To those poor souls that lose their DF save game by playing it in the cloud. They do get a warning. I tested that GFN has a popup that roughly says "Are you sure you want to play this game as it doesn't support Steam Cloud and you may lose your saves?"

**Comment by myk002** [comment]
the `zip` program has the `--filesync` option, which will ensure a zip file is up to date with the original uncompressed files. For directories that have not changed or in which only a few files have changed, it is much faster. I'm not sure if we have access to that functionality from the libzip library we already link to, though

**Comment by myk002** [comment]
some design thoughts (in progress)

### architecture
- cpp plugin to manage file i/o, post-save sync hook, and other business logic
- overlay to add cloud sync interface to DF title screen
- launchdf RPC for Steam communications, e.g. storage limit querying, background network I/O

### data
- dfhack-config/cloud-save.json
- dfhack-cloud-save/*/*.zip

Bulk data is a separate directory from `dfhack-config` since the usage profile is so different. Separating the top-level directories also gives us the flexibility to independently change the cloud sync parameters (e.g. whether to use Steam Auto-Cloud). Finally, it makes it easier for the player if they want to delete the entire directory to save space.

I'm putting `.zip` in as a placeholder. The archive format can be flexible. Whatever is accessible/reliable on all platforms, produces reasonable compression to reduce upload/download times, and does not take too long to compress.

The limits we need to work within are:
- maximum size of a single file: 100M
- maximum number of files across the entire cloud save (including `dfhack-config`): 10,000
- maximum size of backed up data (including `dfhack-config`): 9.7G (10 billion bytes)

Each vanilla savegame consists of:
- one large `world` file (unbounded in size, but 500M for a large world at year 500 is a good estimate)
- one file per site, each of which can be a few megs
- up to thousands of tiny data files representing art, units, etc.

Files for each savegame will be collected in a subdirectory named after that savegame. The limit for a single file backed up to cloud save is 100M, so we will need to split the world files (which can be very large) into smaller chunks. We should choose a small enough chunk size so that it reduces the risk of a failed transfer. All files other than the world file are ridiculously small the most obvious split would be: everything other than `world.*` in a single archive, `world.*` itself in a chunked zip where each chunk is a limited size. we have 9.5G of space to work with and a maximum file count of 10k. I was not able to find whether directories count as files for the purposes of enforcing the limit, so let's assume yes for safety.

Therefore, each save will take:
- 1 file for the directory
- 1 file for the "not world" zip
- world_file_size / size_limit + 1 files for the world chunks

savegames on my drive, many of which were collected from players when investigating bug reports, have world files that range from 260K to 517,000K. The Steam Cloud API has a limit of 100M per file, so let's see where a chunk size of 50M gets us:

on the low end (260K world file):

- 3 files per save (1 for the dir, 1 for the not world zip, 1 for the world zip)
- 3333 savegames (limited by number of files)

on the high end (517M per world file)

- files per save (1 for the dir, 1 for the not world zip, 11 for the world zip)
- 18 savegames (limited by file size)

so we're much more sensitive to file size than file count. we could even lower the chunk size to 5M and be ok (if that's actually beneficial).

however, 18 savegames is not very many. players who run large, older worlds may want more control over which games get backed up. And we certainly need to indicate when limits are being approached/exceeded.

### flows

#### player wants to enable cloud sync
1. player launches DF
2. on title screen, player clicks "Enable cloud save"

#### player wants to disable cloud sync and clear storage

#### player starts game from fully synced state

#### player starts game with newer state on cloud

#### player starts game with newer local state

#### player saves game from within DF

### overlay
- show warning if launchdf is not running
- show warning if enabled and Steam not loaded
- don't show overlay if not enabled and Steam not loaded

### auto cloud save vs. managed
This choice has a significant effect on the UX and implementation complexity. With auto cloud save, you specify a directory in the Steamworks app configuration. The steam client enforces that configuration by copying that directory from the cloud to the local drive when the app starts and copying the changed files in the directory from the local drive to the cloud after the app exits.

Documentation here: https://partner.steamgames.com/doc/features/cloud

Allowing the Steam client to handle syncing means that DFHack doesn't have to do anything with network I/O or conflict resolution. This greatly simplifies the implementation.

However, it also means that there are many sacrifices around UX:
- no background I/O, meaning that synchronizing with the cloud only happens at app exit, which can take a long time. the player must either keep the Steam client untouched while it finishes syncing, or they close it anyway and we fail to sync to the cloud

### conflict resolution

### integration with other tools
- `die`: show warning if data is not yet synchronized with the cloud
